pkg/stub/status.go

```diff
@@ -0,0 +1,196 @@
+
+	if isNotFound {
+		co.Status.Version = operatorversion.Version
+
+		if err := sdk.Create(co); err != nil {
+			logrus.Errorf("syncOperatorStatus: failed to create ClusterOperator %s/%s: %v",
+				co.Namespace, co.Name, err)
+		} else {
+			logrus.Infof("syncOperatorStatus: created ClusterOperator %s/%s (UID %v)",
+				co.Namespace, co.Name, co.UID)
+		}
+
+		return
+	}
+
+	if clusteroperator.ConditionsEqual(oldConditions, co.Status.Conditions) {
+		return
+	}
+
+	if err := sdk.Update(co); err != nil {
+		logrus.Errorf("syncOperatorStatus: failed to update status of ClusterOperator %s/%s: %v",
+			co.Namespace, co.Name, err)
+	}
+}
+
+// getOperatorState gets and returns the resources necessary to compute the
+// operator's current state.
+func (h *Handler) getOperatorState() (*corev1.Namespace, []ingressv1alpha1.ClusterIngress, []appsv1.DaemonSet, error) {
+	ns, err := h.ManifestFactory.RouterNamespace()
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf(
+			"error building router namespace: %v", err)
+	}
+
+	if err := sdk.Get(ns); err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil, nil, nil
+		}
+
+		return nil, nil, nil, fmt.Errorf(
+			"error getting Namespace %s: %v", ns.Name, err)
+	}
+
+	ingressList := &ingressv1alpha1.ClusterIngressList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ClusterIngress",
+			APIVersion: "ingress.openshift.io/v1alpha1",
+		},
+	}
+	err = sdk.List(h.Namespace, ingressList,
+		sdk.WithListOptions(&metav1.ListOptions{}))
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf(
+			"failed to list ClusterIngresses: %v", err)
+	}
+
+	daemonsetList := &appsv1.DaemonSetList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "DaemonSet",
+			APIVersion: "apps/v1",
+		},
+	}
+	err = sdk.List(h.Namespace, daemonsetList,
+		sdk.WithListOptions(&metav1.ListOptions{}))
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf(
+			"failed to list DaemonSets: %v", err)
+	}
+
+	return ns, ingressList.Items, daemonsetList.Items, nil
+}
+
+// computeStatusConditions computes the operator's current state.
+func computeStatusConditions(conditions []osv1.ClusterOperatorStatusCondition, ns *corev1.Namespace, ingresses []ingressv1alpha1.ClusterIngress, daemonsets []appsv1.DaemonSet) []osv1.ClusterOperatorStatusCondition {
+	failingCondition := &osv1.ClusterOperatorStatusCondition{
+		Type:   osv1.OperatorFailing,
+		Status: osv1.ConditionUnknown,
+	}
+	if ns == nil {
+		failingCondition.Status = osv1.ConditionTrue
+		failingCondition.Reason = "NoNamespace"
+		failingCondition.Message = "router namespace does not exist"
+	} else {
+		failingCondition.Status = osv1.ConditionFalse
+	}
+	conditions = clusteroperator.SetStatusCondition(conditions,
+		failingCondition)
+
+	progressingCondition := &osv1.ClusterOperatorStatusCondition{
+		Type:   osv1.OperatorProgressing,
+		Status: osv1.ConditionUnknown,
+	}
+	numIngresses := len(ingresses)
+	numDaemonsets := len(daemonsets)
+	if numIngresses == numDaemonsets {
```

Comment:
When an Ingress CR is deleted, corresponding daemonset may not be deleted yet (pending op) and the length of numIngresses and numDaemonsets will defer, don't we want to say OperatorProgressing = True in this case?